tests/unit/source_code/test_graph.py

```diff
@@ -0,0 +1,183 @@
+    """Register a locatable dependency"""
+    dependency_resolver = DependencyResolver([LocalFileResolver(FileLoader())], mock_path_lookup)
+    graph = DependencyGraph(
+        dependency=file_dependency, parent=None, resolver=dependency_resolver, path_lookup=mock_path_lookup
+    )
+
+    maybe = graph.register_dependency(file_dependency)
+
+    assert len(maybe.problems) == 0
+    assert maybe.graph.dependency == file_dependency
+
+
+def test_dependency_graph_register_new_dependency(mock_path_lookup, file_dependency):
+    """Register a new dependency"""
+    new_dependency = Dependency(FileLoader(), mock_path_lookup.resolve(Path("root1.run.py")))
+    dependency_resolver = DependencyResolver([LocalFileResolver(FileLoader())], mock_path_lookup)
+    graph = DependencyGraph(
+        dependency=file_dependency, parent=None, resolver=dependency_resolver, path_lookup=mock_path_lookup
+    )
+
+    maybe = graph.register_dependency(new_dependency)
+
+    assert len(maybe.problems) == 0
+    assert maybe.graph.dependency == new_dependency
+
+
+def test_dependency_graph_register_new_dependency_with_problem(mock_path_lookup, file_dependency):
+    """Register a new dependency with a problem."""
+    new_dependency = Dependency(FileLoader(), mock_path_lookup.resolve(Path("root10.py")))
+    dependency_resolver = DependencyResolver([LocalFileResolver(FileLoader())], mock_path_lookup)
+    graph = DependencyGraph(
+        dependency=file_dependency, parent=None, resolver=dependency_resolver, path_lookup=mock_path_lookup
+    )
+
+    maybe = graph.register_dependency(new_dependency)
+
+    assert len(maybe.problems) == 1
+    assert maybe.problems[0].source_path == new_dependency.path
+
+
+def test_dependency_resolver_resolve_library_not_found(mock_path_lookup):
+    """Resolve a not found library"""
+    dependency_resolver = DependencyResolver([], mock_path_lookup)
+    maybe = dependency_resolver.resolve_library(mock_path_lookup, "not-found")
+    assert len(maybe.problems) > 0
+    assert maybe.problems[0] == DependencyProblem("library-not-found", "Could not locate library: not-found")
+
+
+def test_dependency_resolver_resolve_library(mock_path_lookup):
+    """Resolve a found library"""
+    file_loader = FileLoader()
+    site_package = SitePackage(Path("test-0.1.0.dist-info/"), ["test"], [Path("test")])
+    container = SitePackageContainer(file_loader, site_package)
+    dependency = Dependency(WrappingLoader(container), Path("pytest"))
+
+    dependency_resolver = DependencyResolver([PipResolver(file_loader)], mock_path_lookup)
+    # TODO: Install pytest when resolving library, prefer to use a resolver without internet access
+    maybe = dependency_resolver.resolve_library(mock_path_lookup, "pytest")
+
+    assert len(maybe.problems) == 0
+    assert maybe.dependency == dependency
+
+
+@pytest.mark.fail("Fails because pytest has a try-except ImportError")
```

Comment:
@ericvergnaud : This is a lovely exception, that we do not handle and which causes problems

``` pytest
__all__ = ["__version__", "version_tuple"]

try:
    from ._version import version as __version__
    from ._version import version_tuple
except ImportError:  # pragma: no cover
    # broken installation, we don't even try
    # unknown only works because we do poor mans version compare
    __version__ = "unknown"
    version_tuple = (0, 0, "unknown")
```